### ISSUE TITLE
typescript doesn't work on svelte reactive variables

### ISSUE THREAD

**Issue by Alexandre-Fernandez** [sveltejs/language-tools#1575]

**Comment by Alexandre-Fernandez** [comment]
when combined with #7754, it gets really annoying since it removes an easy workaround

**Comment by jasonlyu123** [comment]
The example you have is similar to this in plain typescript:

```ts
let foo: string | undefined = undefined

function someFunction(param: string) {
	return param
}

let bar = foo;

() => {
	if (bar) {
		someFunction(foo);
	}

}
```

As you can see in the [typescript playground](https://www.typescriptlang.org/play?#code/DYUwLgBAZg9jBcEDOYBOBLAdgcwgHwgFdMATEKLEEiAXiNPMpIFgAoNqYgYzHRk2QwAtiABi3XvwAUABwCGqOUMQoMOAJQQA3mwCQqcIVQD5ioWwC+bNqEgAjBbWhwA3NdZTNNAHza96KAgpB1RNHVZdXSRhMQk+TClYGHU3CKt2VgsgA). This also has the same error. 

In most cases, Typescript's type guard only narrows down the variable you check, not the one you have assigned with the same value. What's your use case for this? Do you have a more complex case where you can't just test the variable you want to use?

**Comment by Alexandre-Fernandez** [comment]
If I make a reactive variable `$: isLoaded = var1 & var2 & var3` and then wrap the template with `{#if isLoaded}``{/if}`, I'll get undefined errors everytime I'll try to use `var1`, `var2` or `var3` inside the conditional statement.

And since it's not even possible to use the `!` inside the template to tell typescript that it is defined it gets very annoying and the only solution is to prepare all the variables inside the `<script>` tags even when it makes no sense to do so.

**Comment by jasonlyu123** [comment]
> In svelte reactive variables (as far as I know) are always equal to the same thing, when you set $: isLoaded = var1 & var2 & var3 you can't say that it is now equal to var4.

Yes, you can do that. It just makes no sense to do it in most cases. 

Indeed this kind of error is not going to happen in the runtime. But it is more complex for the type checker to check if it's possible. This is also true in TypeScript itself. The control flow analysis has its limitations. 

Anyway, I not trying to say it wrong to make it work. It would be great if it just work. But I can't think of a way to make it work. If you can think of a semantically similar TypeScript code we can see if it could work. Otherwise, I don't think this will be changed anytime soon.
